### PR TITLE
tsdb: Support native histograms in snapshot on shutdown

### LINKED ISSUE #11262: tsdb (histograms): Support native histograms in snapshot on shutdown
See http://ganeshvernekar.com/blog/prometheus-tsdb-snapshot-on-shutdown/ for context on the in-memory snapshot on shutdown.

The snapshot-on-shutdown feature itself is experimental and has to be activated with a feature flag. Therefore, this is only P2 because not having the support only means that you cannot switch on both feature flags at the same time.

Tentatively assigning to @codesome as he is the expert for the snapshot-on-shutdown feature.

### PR DESCRIPTION
**Pull request by marctc**
This PR enables the abiity to store encoded histograms and float histograms in the snapshots.
This is needed when snapshot restoring is enabled during shutdowns.

Fixes #11262

### CODE REVIEW

File: tsdb/head_test.go
Comment (marctc) [inline tsdb/head_test.go:3186]: @codesome I had to use `GaugeHistograms` in the meantime as I'm getting this error when using histograms:
```
 	            	    h: (*histogram.Histogram)({
        	            	-    CounterResetHint: (histogram.CounterResetHint) 2,
        	            	+    CounterResetHint: (histogram.CounterResetHint) 0,
        	            	     Schema: (int32) 1,
```

**Comment by marctc** [comment]
I'll refactor these methods:
 https://github.com/prometheus/prometheus/blob/291ab4d0bc6ea9f0e61073f303bd3982e9727b8d/tsdb/record/record.go#L704
https://github.com/prometheus/prometheus/blob/291ab4d0bc6ea9f0e61073f303bd3982e9727b8d/tsdb/record/record.go#L421

With `EncodeHistogram` and `DecodeHistogram` unless you have a second opinion about it.

File: tsdb/head_test.go
Comment (codesome) [reply tsdb/head_test.go:3186]: Gauge histogram is ok, you can remove this comment. The 2 vs 0 is just the counter reset information. Gauge should make things simpler to test.

File: tsdb/head_wal.go
Comment (codesome) [inline tsdb/head_wal.go:1411]: You missed `lastFloatHistogramValue`

**Review by codesome — COMMENTED** [review]
Things are looking good! Can we move the changes in `model/histogram` into `tsdb/record` as mentioned in the comments? In which case we won't need additional tests for encoding/decoding as well.

File: tsdb/head_wal.go
Comment (codesome) [reply tsdb/head_wal.go:1411]: If no tests failed, I think something is missing in testing?

**Comment by sylr** [comment]
Hey @marctc 👋 

Are you able to move forward with this, it is something I am greatly looking forward!

Thank you!

**Comment by marctc** [comment]
Hey @sylr. Yes I will, I'm currently on vacation. Soon :wink:

**Comment by sylr** [comment]
@codesome any chance you could review the latest changes ? Thank you.

**Comment by beorn7** [comment]
@codesome what are the chances you can look at this any time soon? Otherwise, @jesusvazquez , would you be able to continue the review?

**Review by jesusvazquez — COMMENTED** [review]
I've checked this and it looks good to me. I think there is a meaningful test in head_test.go called `TestSnapshotError` where a few comprobations are made in case the snapshot is corrupted.

I was thinking maybe, for completeness, we could extend this test to add a single histogram/float histogram sample and then make sure that the same expectations the current test has are met.

I'm happy to discuss this if you think it doesnt add that much value :+1:  Once we clarify this I'm happy to approve and merge this unless @codesome has something else to say.

**Review by jesusvazquez — APPROVED** [review]
LGTM!

**Pull request merged**
